File: python/polars_sdist/sampling.py
from __future__ import annotations

import polars as pl

from polars_sdist._polars_sdist import sample_direct as _sample_rs


def _sample(
    dist: str,
    n: int,
    param1: float,
    param2: float | None = None,
    param3: float | None = None,
    seed: int | None = None,
) -> pl.Series:
    """Generate n samples from the given distribution with fixed parameters."""
    return _sample_rs(dist, n, param1, param2, param3, seed)


def sample_normal(n: int, mu: float = 0.0, sigma: float = 1.0, seed: int | None = None) -> pl.Series:
    return _sample("normal", n, mu, sigma, seed=seed)


def sample_lognormal(n: int, mu: float = 0.0, sigma: float = 1.0, seed: int | None = None) -> pl.Series:
    return _sample("lognormal", n, mu, sigma, seed=seed)


def sample_beta(n: int, alpha: float, beta: float, seed: int | None = None) -> pl.Series:
    return _sample("beta", n, alpha, beta, seed=seed)


def sample_gamma(n: int, shape: float, rate: float, seed: int | None = None) -> pl.Series:
    return _sample("gamma", n, shape, rate, seed=seed)


def sample_cauchy(n: int, location: float = 0.0, scale: float = 1.0, seed: int | None = None) -> pl.Series:
    return _sample("cauchy", n, location, scale, seed=seed)


def sample_chi_squared(n: int, df: float, seed: int | None = None) -> pl.Series:
    return _sample("chi_squared", n, df, seed=seed)


def sample_exponential(n: int, lambda_: float, seed: int | None = None) -> pl.Series:
    return _sample("exponential", n, lambda_, seed=seed)


def sample_fisher_snedecor(n: int, d1: float, d2: float, seed: int | None = None) -> pl.Series:
    return _sample("fisher_snedecor", n, d1, d2, seed=seed)


def sample_gumbel(n: int, location: float = 0.0, scale: float = 1.0, seed: int | None = None) -> pl.Series:
    return _sample("gumbel", n, location, scale, seed=seed)


def sample_inverse_gamma(n: int, shape: float, scale: float, seed: int | None = None) -> pl.Series:
    return _sample("inverse_gamma", n, shape, scale, seed=seed)


def sample_laplace(n: int, location: float = 0.0, scale: float = 1.0, seed: int | None = None) -> pl.Series:
    return _sample("laplace", n, location, scale, seed=seed)


def sample_pareto(n: int, shape: float, scale: float, seed: int | None = None) -> pl.Series:
    return _sample("pareto", n, shape, scale, seed=seed)


def sample_students_t(n: int, df: float, seed: int | None = None) -> pl.Series:
    return _sample("students_t", n, df, seed=seed)


def sample_triangular(n: int, min: float, max: float, mode: float, seed: int | None = None) -> pl.Series:
    return _sample("triangular", n, min, max, mode, seed=seed)


def sample_uniform(n: int, a: float = 0.0, b: float = 1.0, seed: int | None = None) -> pl.Series:
    return _sample("uniform", n, a, b, seed=seed)


def sample_weibull(n: int, shape: float, scale: float, seed: int | None = None) -> pl.Series:
    return _sample("weibull", n, shape, scale, seed=seed)


def sample_pert(n: int, min: float, max: float, mode: float, seed: int | None = None) -> pl.Series:
    return _sample("pert", n, min, max, mode, seed=seed)


def sample_skew_normal(n: int, location: float, scale: float, shape: float, seed: int | None = None) -> pl.Series:
    return _sample("skew_normal", n, location, scale, shape, seed=seed)


def sample_inverse_gaussian(n: int, mean: float, shape: float, seed: int | None = None) -> pl.Series:
    return _sample("inverse_gaussian", n, mean, shape, seed=seed)


def sample_frechet(n: int, location: float, scale: float, shape: float, seed: int | None = None) -> pl.Series:
    return _sample("frechet", n, location, scale, shape, seed=seed)


def sample_zeta(n: int, s: float, seed: int | None = None) -> pl.Series:
    return _sample("zeta", n, s, seed=seed)


def sample_zipf(n: int, n_elements: int, s: float, seed: int | None = None) -> pl.Series:
    return _sample("zipf", n, float(n_elements), s, seed=seed)


def sample_bernoulli(n: int, p: float, seed: int | None = None) -> pl.Series:
    return _sample("bernoulli", n, p, seed=seed)


def sample_binomial(n: int, trials: int, p: float, seed: int | None = None) -> pl.Series:
    return _sample("binomial", n, float(trials), p, seed=seed)


def sample_geometric(n: int, p: float, seed: int | None = None) -> pl.Series:
    return _sample("geometric", n, p, seed=seed)


def sample_poisson(n: int, lambda_: float, seed: int | None = None) -> pl.Series:
    return _sample("poisson", n, lambda_, seed=seed)

File: src/distributions/sampling.rs
use polars::prelude::*;
use pyo3_polars::derive::polars_expr;
use rand::prelude::*;

use super::DistKwargs;
use crate::utils::make_rng;

/// Macro to construct a rand_distr distribution and sample N times into Vec<f64>.
macro_rules! dispatch_sample {
    ($name:expr, $p1:expr, $p2:expr, $p3:expr, $n:expr, $rng:expr) => {{
        macro_rules! err {
            ($dist_name:expr, $e:expr) => {
                PolarsError::ComputeError(format!("{}: {}", $dist_name, $e).into())
            };
        }
        match $name {
            "normal" => {
                let d = rand_distr::Normal::new($p1, $p2.unwrap()).map_err(|e| err!("normal", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "lognormal" => {
                let d = rand_distr::LogNormal::new($p1, $p2.unwrap()).map_err(|e| err!("lognormal", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "beta" => {
                let d = rand_distr::Beta::new($p1, $p2.unwrap()).map_err(|e| err!("beta", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "gamma" => {
                // statrs Gamma(shape, rate) but rand_distr Gamma(shape, scale=1/rate)
                let d = rand_distr::Gamma::new($p1, 1.0 / $p2.unwrap()).map_err(|e| err!("gamma", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "cauchy" => {
                let d = rand_distr::Cauchy::new($p1, $p2.unwrap()).map_err(|e| err!("cauchy", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "chi_squared" => {
                let d = rand_distr::ChiSquared::new($p1).map_err(|e| err!("chi_squared", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "exponential" => {
                let d = rand_distr::Exp::new($p1).map_err(|e| err!("exponential", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "fisher_snedecor" => {
                let d = rand_distr::FisherF::new($p1, $p2.unwrap()).map_err(|e| err!("fisher_snedecor", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "gumbel" => {
                let d = rand_distr::Gumbel::new($p1, $p2.unwrap()).map_err(|e| err!("gumbel", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "inverse_gamma" => {
                // Sample as 1/Gamma(alpha, 1/beta)
                let d = rand_distr::Gamma::new($p1, 1.0 / $p2.unwrap()).map_err(|e| err!("inverse_gamma", e))?;
                Ok((0..$n).map(|_| 1.0 / $rng.sample(d)).collect::<Vec<f64>>())
            }
            "laplace" => {
                // Laplace via inverse CDF
                let uniform = rand_distr::Uniform::new(0.0f64, 1.0).unwrap();
                let loc = $p1;
                let scale = $p2.unwrap();
                Ok((0..$n).map(|_| {
                    let u: f64 = $rng.sample(uniform);
                    let v = u - 0.5;
                    loc - scale * v.signum() * (1.0 - 2.0 * v.abs()).ln()
                }).collect::<Vec<f64>>())
            }
            "pareto" => {
                // rand_distr::Pareto::new(scale, shape)
                let d = rand_distr::Pareto::new($p2.unwrap(), $p1).map_err(|e| err!("pareto", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "students_t" => {
                let d = rand_distr::StudentT::new($p1).map_err(|e| err!("students_t", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "triangular" => {
                let d = rand_distr::Triangular::new($p1, $p2.unwrap(), $p3.unwrap()).map_err(|e| err!("triangular", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "uniform" => {
                let d = rand_distr::Uniform::new($p1, $p2.unwrap()).map_err(|e| err!("uniform", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "weibull" => {
                // rand_distr::Weibull::new(scale, shape)
                let d = rand_distr::Weibull::new($p2.unwrap(), $p1).map_err(|e| err!("weibull", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "pert" => {
                // Pert::new(min, max).with_mode(mode)
                let d = rand_distr::Pert::new($p1, $p2.unwrap())
                    .with_mode($p3.unwrap())
                    .map_err(|e| err!("pert", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "skew_normal" => {
                let d = rand_distr::SkewNormal::new($p1, $p2.unwrap(), $p3.unwrap()).map_err(|e| err!("skew_normal", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "inverse_gaussian" => {
                let d = rand_distr::InverseGaussian::new($p1, $p2.unwrap()).map_err(|e| err!("inverse_gaussian", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "frechet" => {
                // Frechet::new(location, scale, shape)
                let d = rand_distr::Frechet::new($p1, $p2.unwrap(), $p3.unwrap()).map_err(|e| err!("frechet", e))?;
                Ok((0..$n).map(|_| $rng.sample(d)).collect::<Vec<f64>>())
            }
            "zeta" => {
                let d = rand_distr::Zeta::new($p1).map_err(|e| err!("zeta", e))?;
                Ok((0..$n).map(|_| $rng.sample::<f64, _>(&d)).collect::<Vec<f64>>())
            }
            "zipf" => {
                // Zipf::new(n, s) where n is f64
                let d = rand_distr::Zipf::new($p1, $p2.unwrap()).map_err(|e| err!("zipf", e))?;
                Ok((0..$n).map(|_| $rng.sample::<f64, _>(&d)).collect::<Vec<f64>>())
            }
            "bernoulli" => {
                let d = rand_distr::Bernoulli::new($p1).map_err(|e| err!("bernoulli", e))?;
                Ok((0..$n).map(|_| if $rng.sample(&d) { 1.0 } else { 0.0 }).collect::<Vec<f64>>())
            }
            "binomial" => {
                let d = rand_distr::Binomial::new($p1 as u64, $p2.unwrap()).map_err(|e| err!("binomial", e))?;
                Ok((0..$n).map(|_| $rng.sample(&d) as f64).collect::<Vec<f64>>())
            }
            "geometric" => {
                let d = rand_distr::Geometric::new($p1).map_err(|e| err!("geometric", e))?;
                Ok((0..$n).map(|_| $rng.sample(&d) as f64).collect::<Vec<f64>>())
            }
            "poisson" => {
                let d = rand_distr::Poisson::new($p1).map_err(|e| err!("poisson", e))?;
                Ok((0..$n).map(|_| $rng.sample::<f64, _>(&d)).collect::<Vec<f64>>())
            }
            other => {
                Err(PolarsError::ComputeError(format!("unknown distribution for sampling: {other}").into()))
            }
        }
    }};
}

/// Core sampling logic: returns a Vec<f64> of N draws from the given distribution.
pub(crate) fn sample_vec(
    dist: &str,
    n: usize,
    param1: f64,
    param2: Option<f64>,
    param3: Option<f64>,
    seed: Option<u64>,
) -> PolarsResult<Vec<f64>> {
    let mut rng = make_rng(seed);
    dispatch_sample!(dist, param1, param2, param3, n, rng)
}

/// Fixed-parameter sampling: generates a series of N draws.
#[polars_expr(output_type=Float64)]
fn dist_sample(inputs: &[Series], kwargs: DistKwargs) -> PolarsResult<Series> {
    let n = inputs[0].len();
    let values = sample_vec(
        kwargs.dist.as_str(),
        n,
        kwargs.param1,
        kwargs.param2,
        kwargs.param3,
        kwargs.seed,
    )?;
    let ca = Float64Chunked::from_vec("sample".into(), values);
    Ok(ca.into_series())
}

/// Column-parameterized sampling: per-row distribution params from columns.
#[polars_expr(output_type=Float64)]
fn dist_sample_col(inputs: &[Series], kwargs: DistKwargs) -> PolarsResult<Series> {
    let p1_ca = inputs[0].f64()?;
    let p2_ca = if inputs.len() > 1 {
        Some(inputs[1].f64()?)
    } else {
        None
    };
    let p3_ca = if inputs.len() > 2 {
        Some(inputs[2].f64()?)
    } else {
        None
    };
    let n = p1_ca.len();
    let mut rng = make_rng(kwargs.seed);
    let dist_name = kwargs.dist.as_str();

    let mut values: Vec<Option<f64>> = Vec::with_capacity(n);

    for i in 0..n {
        let p1 = p1_ca.get(i);
        let p2 = p2_ca.as_ref().and_then(|c| c.get(i));
        let p3 = p3_ca.as_ref().and_then(|c| c.get(i));

        if p1.is_none() {
            values.push(None);
            continue;
        }
        let p1 = p1.unwrap();

        let sampled: Vec<f64> = dispatch_sample!(dist_name, p1, p2, p3, 1usize, rng)?;
        values.push(Some(sampled[0]));
    }

    let ca = Float64Chunked::from_iter_options("sample".into(), values.into_iter());
    Ok(ca.into_series())
}
